src/ExportHistory/Orchestrations/ExportJobOrchestrator.cs

```diff
@@ -0,0 +1,376 @@
+
+            while (true)
+            {
+                processedCycles++;
+                if (processedCycles > ContinueAsNewFrequency)
+                {
+                    context.ContinueAsNew(new ExportJobRunRequest(input.JobEntityId, ProcessedCycles: 0));
+                    return null!;
+                }
+
+                // Check if job is still active (entity might have been deleted or failed)
+                ExportJobState? currentState = await context.Entities.CallEntityAsync<ExportJobState?>(
+                    input.JobEntityId,
+                    nameof(ExportJob.Get),
+                    null);
+
+                if (currentState == null ||
+                    currentState.Config == null ||
+                    currentState.Status != ExportJobStatus.Active)
+                {
+                    logger.ExportJobOperationWarning(jobId, nameof(ExportJobOrchestrator), "Job is no longer active");
+                    return null;
+                }
+
+                // Call activity to list terminal instances with only necessary information
+                ListTerminalInstancesRequest listRequest = new ListTerminalInstancesRequest(
+                    CompletedTimeFrom: currentState.Config.Filter.CompletedTimeFrom,
+                    CompletedTimeTo: currentState.Config.Filter.CompletedTimeTo,
+                    RuntimeStatus: currentState.Config.Filter.RuntimeStatus,
+                    LastInstanceKey: currentState.Checkpoint?.LastInstanceKey,
+                    MaxInstancesPerBatch: currentState.Config.MaxInstancesPerBatch);
+
+                InstancePage pageResult = await context.CallActivityAsync<InstancePage>(
+                    nameof(ListTerminalInstancesActivity),
+                    listRequest);
+
+                List<string> instancesToExport = pageResult.InstanceIds;
+                long scannedCount = instancesToExport.Count;
+
+                if (scannedCount == 0)
+                {
+                    if (config.Mode == ExportMode.Continuous)
+                    {
+                        logger.ExportJobOperationInfo(jobId, nameof(ExportJobOrchestrator), "No more instances to export - export complete.");
+                        await context.CreateTimer(ContinuousExportIdleDelay, default);
+                        continue;
+                    }
+                    else if (config.Mode == ExportMode.Batch)
+                    {
+                        logger.ExportJobOperationInfo(jobId, nameof(ExportJobOrchestrator), "No more instances to export currently - will retry later.");
+                        break;
+                    }
+                    else
+                    {
+                        throw new InvalidOperationException("Invalid export mode.");
+                    }
+                }
+
+                // Process batch with retry logic
+                BatchExportResult batchResult = await this.ProcessBatchWithRetryAsync(
```

Comment:
Should stuff like this be an activity for additional checkpointing? Same with the commit.

Reply:
its fanning out to activities per instance export inside this function, this function itself(wrapping fan out code) does not need to be in an activity because the fanned out activities already do the checkpointing, even it rerun the wrapping code, it will skip the exported activities. the wrapping code itself is trivial to be checkpointed and we cant fan out activities inside a parent activity